### PR TITLE
:seedling: TK-4853 Add support for providing preflight inputs through a yaml file

### PR DESCRIPTION
**Pull request by prasad-trilio**
Able to read input for preflight run and cleanup through a yaml file.
Can override the inputs of the file through CLI flags.

Added 3 flags of resources requirements during preflight run:
1. --requests: user can specify cpu and memory requests for all pods
2. --limits: user can specify cpu and memory limits for all pods
3. --pvc-storage-request: PVC storage requirement request for volume snapshot check.

Integration tests are added.

### CODE REVIEW

File: tools/preflight/preflight.go [hunk 1]
```diff
@@ -446,6 +453,8 @@ func (o *Options) checkVolumeSnapshot(ctx context.Context) error {
 		waitOptions *wait.PodWaitOptions
 		err         error
 	)
+	updateVolSnapPodResourceReqs(o.PodResourceRequirements)
+	logVolumeSnapshotResReqs(o.Logger)
```
Comment (msurudwad) [inline tools/preflight/preflight.go:457]: instead such log functions.. print whole PreflightOptions struct in json format before running preflight check.
do the same for cleanup command

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: cmd/preflight/cmd/helper.go [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"sigs.k8s.io/yaml"
 )
 
-func setupLogger(logFilePrefix string) error {
+type preflightCmdOps struct {
+	PreflightOps preflight.Options        `json:"preflightOptions"`
+	CleanupOps   preflight.CleanupOptions `json:"cleanupOptions"`
+}
```
[the remark this answers is not in the document]
Comment (prasad-trilio) [reply cmd/preflight/cmd/helper.go:27]: The struct is unexported as we don't want the struct to be used outside preflight cmd package.
The fields are exported to read data from file into the fields.

File: tools/preflight/cleanup.go [hunk 3]
```diff
@@ -13,6 +13,8 @@ import (
 
 type CleanupOptions struct {
 	CommonOptions
+	CleanupMode string `json:"cleanupMode"`
```
[the remark this answers is not in the document]
Comment (prasad-trilio) [reply tools/preflight/cleanup.go:16]: Determines whether the cleanup is to be done for all preflight resources or for a particular preflight uid
if mode is 'all' then clean all resources
if mode is 'uid' then a uid field must be specified in yaml input file.

For CLI, there is no need to specify cleanup-mode. if --uid flag is specified then, mode is uid, all otherwise

File: cmd/preflight/cmd/helper.go [hunk 4]
```diff
@@ -1,17 +1,32 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"sigs.k8s.io/yaml"
 )
 
-func setupLogger(logFilePrefix string) error {
+type preflightCmdOps struct {
+	PreflightOps preflight.Options        `json:"preflightOptions"`
+	CleanupOps   preflight.CleanupOptions `json:"cleanupOptions"`
```
[the remark this answers is not in the document]
Comment (prasad-trilio) [reply cmd/preflight/cmd/helper.go:26]: The purpose of `preflightCmdOps` is to provide interface to run and cleanup commands.
If CommonOptions is added here then cmdOps struct instance must be passed down to the cleanup and run preflight checks funcs.

CommonOptions struct embedded in Run and Cleanup structs seems to be better structure.
@msurudwad please let me know your thoughts on this.

File: cmd/preflight/cmd/run.go [hunk 5]
```diff
@@ -29,9 +29,9 @@ var runCmd = &cobra.Command{
   # run preflight checks with a particular log level
   kubectl tvk-preflight run --storage-class <storage-class-name> --log-level <log-level>
 
-  # cleanup the resources generated during preflight check if preflight check fails. Default is false.
+  # Cleanup the resources generated during preflight check if preflight check fails. Default is false.
   # If the preflight check is successful, then all resources are cleaned.
-  kubectl tvk-preflight run --storage-class <storage-class-name> --cleanup-on-failure 
+  kubectl tvk-preflight run --storage-class <storage-class-name> --Cleanup-on-failure
```
Comment (msurudwad) [inline cmd/preflight/cmd/run.go:34]: revert this change

File: docs/preflight/sample_file.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+run:
+  namespace: default
+  storageClass: csi-gce-pd
+  cleanupOnFailure: true
+  pvcStorageRequest: 2Gi
```
Comment (msurudwad) [inline docs/preflight/sample_file.yaml:5]: keep it 1Gi

File: docs/preflight/sample_file.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+run:
+  namespace: default
+  storageClass: csi-gce-pd
```
Comment (msurudwad) [inline docs/preflight/sample_file.yaml:3]:
```suggestion
  storageClass: default
```

File: tests/preflight/preflight_suite_test.go [hunk 8]
```diff
@@ -60,24 +64,23 @@ const (
...
+	localRegistryFlag     = "--local-registry"
+	serviceAccountFlag    = "--service-account"
+	cleanupOnFailureFlag  = "--cleanup-on-failure"
+	namespaceFlag         = "--namespace"
+	kubeconfigFlag        = "--kubeconfig"
+	logLevelFlag          = "--log-level"
+	configFileFlag        = "--config-file"
+	pvcStorageRequestFlag = "--pvc-storage-request"
+	limitsFlag            = "--limits"
+	requestsFlag          = "--requests"
```
Comment (msurudwad) [inline tests/preflight/preflight_suite_test.go:83]: use flag names from pre-declared consts

File: tests/preflight/preflight_suite_test.go [hunk 9]
```diff
@@ -109,6 +117,17 @@ var (
...
 
+	resourceReqs = corev1.ResourceRequirements{
+		Requests: map[corev1.ResourceName]resource.Quantity{
+			corev1.ResourceMemory: resource.MustParse("64Mi"),
+			corev1.ResourceCPU:    resource.MustParse("250m"),
+		},
+		Limits: map[corev1.ResourceName]resource.Quantity{
+			corev1.ResourceMemory: resource.MustParse("128Mi"),
+			corev1.ResourceCPU:    resource.MustParse("500m"),
+		},
```
Comment (msurudwad) [inline tests/preflight/preflight_suite_test.go:128]: there're consts declared for resource values 64Mi, etc.. use those

File: tools/preflight/cleanup.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (msurudwad) [reply tools/preflight/cleanup.go:15]: ok, why can't we keep the same behavior as CLI flag? if `uid` value is empty or not given in yaml file, cleanup all resources. If uid value is given, cleanup that uid resources only.

File: tools/preflight/preflight.go [hunk 10]
```diff
@@ -16,22 +16,28 @@ import (
...
-type Options struct {
+// RunOptions input options required for running preflight.
+type RunOptions struct {
+	StorageClass                string            `json:"storageClass"`
+	SnapshotClass               string            `json:"snapshotClass"`
+	LocalRegistry               string            `json:"localRegistry"`
+	ImagePullSecret             string            `json:"imagePullSecret"`
+	ServiceAccountName          string            `json:"serviceAccount"`
+	PerformCleanupOnFail        bool              `json:"cleanupOnFailure"`
+	PVCStorageRequest           resource.Quantity `json:"pvcStorageRequest"`
```
Comment (msurudwad) [inline tools/preflight/preflight.go:34]: add `omitempty` to json tag

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: cmd/preflight/cmd/helper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (msurudwad) [reply cmd/preflight/cmd/helper.go]: If CommonOptions are kept in `preflightCmdOps` struct then, users won't need to specify same config twice in yaml file. At present, users will need to specify common values in both `run` and `cleanup` section. This can user experience can be enhanced.
what is the challenge in doing this `If CommonOptions is added here then cmdOps struct instance must be passed down to the cleanup and run preflight checks funcs`?

[the remark this answers is not in the document]
Comment (msurudwad) [reply cmd/preflight/cmd/helper.go]: This is an enhancement, no hard requirement to do so.

[the remark this answers is not in the document]
Comment (prasad-trilio) [reply cmd/preflight/cmd/helper.go]: Will take up this task in a separate ticket

**Review by msurudwad — APPROVED** [review]
